pkg/virtctl/guestfs/guestfs.go

```diff
@@ -0,0 +1,448 @@
+	cmd.PersistentFlags().StringVar(&namespace, "pvc-namespace", "default", "namespace of the pvc")
+	cmd.PersistentFlags().BoolVar(&kvm, "kvm", true, "Use kvm for the libguestfs-tools container")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+func usage() string {
+	usage := `  # Create a pod with libguestfs-tools, mount the pvc and attach a shell to it:
+  {{ProgramName}} guestfs --pvc pvc`
+	return usage
+}
+
+// ClientCreator is a function to return the Kubernetes client
+type ClientCreator func(config *rest.Config, virtClientConfig clientcmd.ClientConfig) (*K8sClient, error)
+
+var createClientFunc ClientCreator
+
+// SetClient allows overriding the default Kubernetes client. Useful for creating a mock function for the testing.
+func SetClient(f ClientCreator) {
+	createClientFunc = f
+}
+
+// SetDefaulClient sets the default function to create the Kubernetes client
+func SetDefaulClient() {
+	createClientFunc = createClient
+}
+
+// AttacherCreator is a function that attach a command to a pod using the Kubernetes client
+type AttacherCreator func(client *K8sClient, p *corev1.Pod, command string) error
+
+var createAttacherFunc AttacherCreator
+
+// SetAttacher allows overriding the default attacher function. Useful for creating a mock function for the testing.
+func SetAttacher(f AttacherCreator) {
+	createAttacherFunc = f
+}
+
+// SetDefaulAttacher sets the default function to attach to a pod
+func SetDefaulAttacher() {
+	createAttacherFunc = createAttacher
+}
+
+func init() {
+	SetDefaulClient()
+	SetDefaulAttacher()
+}
+
+func (c *guestfsCommand) run(cmd *cobra.Command, args []string) error {
+	var inUse bool
+	conf, err := c.clientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+	client, err := createClientFunc(conf, c.clientConfig)
+	if err != nil {
+		return err
+	}
+	// if the image has not been set from the command line, then we fetch the registry and the image information from the kubevirt CRD
+	if image == "" {
+		if err = client.setImageFromKubeVirt(); err != nil {
+			return err
+		}
+	}
+	exist, _ := client.existsPVC(pvc, namespace)
+	if !exist {
+		return fmt.Errorf("The PVC %s doesn't exist", pvc)
+	}
+	inUse, err = client.isPVCinUse(pvc, namespace)
+	if err != nil {
+		return err
+	}
+	if inUse {
+		return fmt.Errorf("PVC %s is used by another pod", pvc)
+	}
+	isBlock, err := client.isPVCVolumeBlock(pvc, namespace)
+	if err != nil {
+		return err
+	}
+	defer client.removePod(namespace)
+	return client.createInteractivePodWithPVC(pvc, image, namespace, "/entrypoint.sh", []string{}, isBlock)
+}
+
+// K8sClient holds the information of the Kubernetes client
+type K8sClient struct {
+	Client     kubernetes.Interface
+	config     *rest.Config
+	VirtClient kubecli.KubevirtClient
+}
+
+func (client *K8sClient) setImageFromKubeVirt() error {
+	kubevirt, err := client.VirtClient.KubeVirt(kubevirtNs).Get(kubevirtCRD, &metav1.GetOptions{})
```

Comment:
Do "regular" users (say admin for one namespace) have permission to get the KubeVirt CRD? 

All authenticated users can get CDIConfig, but is that also true here as well?  If not, I suspect only cluster admins will be able to succeed here.

Reply:
@mhenriks good point...I have to check the RBAC for the CRD but it's a very good (and painfull!) point

Reply:
a normal user can always overwrite the default image with `--image`, and then the setup works. However, it will be nice to give some hints also for standard users

Reply:
I tried to create a user that can get resources only from her namespace. It is true that I'm not able to get the kubevirt crd but I cannot also create any vms:
```bash
$ virtctl guestfs --pvc pvc-test
cannot obtain KubeVirt resource: kubevirts.kubevirt.io "kubevirt" is forbidden: User "system:serviceaccount:mynamespace:mynamespace-user" cannot get resource "kubevirts" in API group "kubevirt.io" in the namespace "kubevirt"
$ k apply -f examples/vmi-ephemeral.yaml 
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "kubevirt.io/v1, Resource=virtualmachineinstances", GroupVersionKind: "kubevirt.io/v1, Kind=VirtualMachineInstance"
Name: "vmi-ephemeral", Namespace: "mynamespace"
from server for: "examples/vmi-ephemeral.yaml": virtualmachineinstances.kubevirt.io "vmi-ephemeral" is forbidden: User "system:serviceaccount:mynamespace:mynamespace-user" cannot get resource "virtualmachineinstances" in API group "kubevirt.io" in the namespace "mynamespace"

```

Reply:
@mhenriks you are right for user that cannot get kubevirt resource, the command doesn't work. The difference between cdiconfig and kubevirt crd is that cdiconfig is not namespaced and kubevirt crd is

Reply:
the only possible alternative I can figure out is to add an additional crd not namepsaced that contains the version, registry and pull policy information. It just complicates a bit further this PR

Reply:
It's more about RBAC than namespaced vs non namespaced.  But eliminating the need to know the install namespace is also helpful

Reply:
Another alternative is to create the libguestfs pods server side.  Would require something like a libguestfs CRD.  virtctl would create the CR, a controller would create the pod and set the name of the pod in the status of the CR.  Would also have to make sure that user has permission to create pods in the target namespace via webhook validation.

Reply:
To address what Michael pointed out. It is possible to create a CRD that holds the image information like:
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: gsconfigs.libguestfs.kubevirt.io
spec:
  group: libguestfs.kubevirt.io
  scope: Cluster
  names:
    listKind: GuestfsImageConfigList
    plural: gsconfigs
    singular: gsconfig
    kind: GuestfsImageConfig
    shortNames:
    - gsconfig
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          description: GuestfsImageConfig contains the information for the libguestfs-image used by virtctl guestfs command
          type: object
          properties:
            spec:
              type: object
              properties:
                registry:
                  type: string
                image:
                  type: string
                tag:
                  type: string
                imagePullPolicy:
                  description: PullPolicy describes a policy for if/when to pull a container image
                  enum:
                  - Always
                  - IfNotPresent
                  - Never
                  type: string
```
and add `get,list,watch` access to all the service account and authenticated user like:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: libguestfs.kubevirt.io:config-reader
rules:
- apiGroups:
  - libguestfs.kubevirt.io
  resources:
  - gsconfigs
  verbs:
  - get
  - list
  - watch
--- 
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: libguestfs.kubevirt.io:config-reader
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: libguestfs.kubevirt.io:config-reader
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:authenticated
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:serviceaccount
```
After it is possible to create the config:
```yaml
apiVersion: "libguestfs.kubevirt.io/v1"
kind: GuestfsImageConfig
metadata:
  name: config
spec:
  registry: "registry:5000"
  image: "libguestfs-tools"
  imagePullPolicy: IfNotPresent
  tag: "devel"
```
After the gsconfig can be getted from any authenticated users/sas:
```bash
kubectl  --context=test-user get gsconfig config -o yaml
apiVersion: libguestfs.kubevirt.io/v1
kind: GuestfsImageConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"libguestfs.kubevirt.io/v1","kind":"GuestfsImageConfig","metadata":{"annotations":{},"name":"config"},"spec":{"image":"libguestfs-tools","imagePullPolicy":"IfNotPresent","registry":"registry:5000","tag":"devel"}}
  creationTimestamp: "2021-06-14T10:25:56Z"
  generation: 1
  name: config
  resourceVersion: "15587"
  selfLink: /apis/libguestfs.kubevirt.io/v1/gsconfigs/config
  uid: 4da2dbeb-bf14-452e-953b-7b0715b32ae7
spec:
  image: libguestfs-tools
  imagePullPolicy: IfNotPresent
  registry: registry:5000
  tag: devel
```

However, this additonal crd and rbac need to be added by the kubevirt operator. Any thought against this approch?

Reply:
Maybe even a separate configuration that simplify the deployment and separate the kubevirt core of this feature

Reply:
I'm playing with the custom crd and it looks nice. However, the integration is pretty huge. I'm thinking to add it in a later PR and use here the hardcoded version. Is it a reasonable approch?

Reply:
hi @rmohr @davidvossel 

I'd like to have some feedback from you if a new crd that holds the container image information is fine. This solves the issues reported by Michael. In the previous version, I was getting registry, tag, and pull policy from the KubeVirt crd, but this doesn't work with users that are not allowed to get kubevirt crd.

Reply:
The approach sounds fine to me in general. Would it make sense to create a `kubevirt-public` namespace and make the CRD namespaced? Also would it make sense to have a common "public" config insteada of a guestfs specific one? Similar to our kubevirt CR, except that it only holds public information?

Reply:
if we're just talking about a way to make some sort of info about the install publicly available, this seems like a similar issue that was already solved by the `version` endpoint.

everyone has permission to view the version endpoint, which the client can retrieve server side version info with like this.

```
$ virtctl version
Client Version: version.Info{GitVersion:"v0.42.0-169+744370164f8e26-dirty", GitCommit:"744370164f8e26ac37b7e32c5bfb19ba114e576f", GitTreeState:"dirty", BuildDate:"2021-06-17T15:01:39Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v0.42.0-169+744370164f8e26-dirty", GitCommit:"744370164f8e26ac37b7e32c5bfb19ba114e576f", GitTreeState:"dirty", BuildDate:"2021-06-17T15:01:39Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
```

Would it be a crazy idea to reuse this version endpoint to pass the guest appliance info or create a new endpoint just for passing the image back?

Reply:
My current problem with that approach is that when I compile the virt-api, I already need to know the release version. For example, I don't know the shas of the image. @davidvossel do you have any suggestions what I could use as tag of the image in the version?

Reply:
after some offline discussion with @rmohr. We might want to expose the guestfs information through another endpoint, and the information exposed there will be populated by virt-api that has access to the KubeVirt CR. Once this is available the information get be retrieved by virtctl

Reply:
Is it fine to integrate the versioning in a second PR and keep in this one the hardcoded version?